File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.cpp
class Solution {
public:
    int furthestBuilding(vector<int>& heights, int bricks, int ladders) {
        
        
        
        priority_queue<int> maxHeap;
        
        for(int i=1; i<heights.size(); i++){
            
            int diff = heights[i]-heights[i-1];
            if(diff<=0)
                continue;
            
            
            bricks -=diff;
            maxHeap.push(diff);
            
            if(bricks<0){
                if(ladders==0){
                    return i-1;
                }else{
                    ladders--;
                    bricks+= maxHeap.top();
                    maxHeap.pop();
                }
            }
        }//for
        
        
        return heights.size()-1;
    }
};

File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.py
class Solution:
    def furthestBuilding(self, heights: List[int], bricks: int, ladders: int) -> int:
        
        maxHeap =[]
        
        
        for i in range(1, len(heights)):
            
            
            diff = heights[i]-heights[i-1]
            
            if(diff<=0):
                continue
                
                
            bricks -=diff
            heapq.heappush(maxHeap, -diff)
            
            
            if bricks <0:
                if ladders ==0:
                    return i-1
                
                ladders -=1
                bricks +=   -heapq.heappop(maxHeap)
                
                
        return len(heights)-1